### PR TITLE
JAMES-3438: Email/set create htmlBody

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSet.scala [hunk 1]
```diff
@@ -48,13 +48,32 @@ object EmailSet {
...
+      builder.setBody(maybeHtmlBody.getOrElse(""), "html", StandardCharsets.UTF_8)
+      builder.build()
+    })
+
+
+  def validateHtmlBody: Either[IllegalArgumentException, Option[String]] = htmlBody match {
+    case None => Right(None)
+    case Some(html :: Nil) => bodyValues.getOrElse(Map()).get(html.partId)
+      .map(part => Right(Some(part.value)))
+      .getOrElse(Left(new IllegalArgumentException("todo")))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSet.scala:76]: case _ => reject (max 1 HTML part)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSet.scala [hunk 2]
```diff
@@ -48,13 +48,32 @@ object EmailSet {
...
-    builder.setBody("", StandardCharsets.UTF_8)
-    builder.build()
+                                subject: Option[Subject],
+                                htmlBody: Option[List[ClientHtmlBody]],
+                                bodyValues: Option[Map[ClientPartId, ClientEmailBodyValue]]) {
+  def toMime4JMessage: Either[IllegalArgumentException, Message] =
+    validateHtmlBody.map(maybeHtmlBody => {
+      val builder = Message.Builder.of
+      subject.foreach(value => builder.setSubject(value.value))
+      builder.setBody(maybeHtmlBody.getOrElse(""), "html", StandardCharsets.UTF_8)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSet.scala:67]: WHat is "html" here?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSetMethod.scala [hunk 3]
```diff
@@ -227,13 +223,18 @@ class EmailSetMethod @Inject()(serializer: EmailSetSerializer,
...
       SMono.fromCallable[CreationResult](() => {
-        val mailboxId: MailboxId = request.mailboxIds.value.headOption.get
-        val appendResult = mailboxManager.getMailbox(mailboxId, mailboxSession)
-          .appendMessage(AppendCommand.builder()
-            .recent()
-            .build(request.toMime4JMessage),
-            mailboxSession)
-        CreationSuccess(clientId, EmailCreationResponse(appendResult.getId.getMessageId))
+        request.toMime4JMessage
+          .fold(e => CreationFailure(clientId, e),
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSetMethod.scala:227]: I would swtich the mono & the fold:

```
request.toMime4JMessage
   .fold(e => SMTP.just(CreationFailure(clientId, e)),
   mime4JMessage => SMono.fromCallable[CreationResult](() => ...)));
```

(validate data before computations)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSetMethod.scala [hunk 4]
```diff
@@ -227,13 +223,18 @@ class EmailSetMethod @Inject()(serializer: EmailSetSerializer,
...
+            message => {
+              val mailboxId: MailboxId = request.mailboxIds.value.headOption.get
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSetMethod.scala:229]: mailboxId don't need to be in the lambda, no?

**Review by chibenwa — COMMENTED** [review]
A rebase to do,

toms of tests to add.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSet.scala (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSet.scala:94]: Can we get a htmlSubType val here?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSetMethodContract.scala [hunk 5]
```diff
@@ -850,6 +850,626 @@ trait EmailSetMethodContract {
...
+      .inPath("methodResponses[0][1].notCreated.aaaaaa")
+      .isEqualTo(
+        s"""{
+           |  "type": "invalidArguments",
+           |  "description": "Expecting htmlBody to contains only 1 part"
+           |}""".stripMargin)
+  }
+
+  @Test
+  def createShouldFailWhenWrongBodyPart(server: GuiceJamesServer): Unit = {
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSetMethodContract.scala:1075]: createShouldFailWhenPartIdMisMatch

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
https://issues.apache.org/jira/browse/JAMES-3438

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**